// live-mutex/dist/index.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
const lmUtils = require("./utils");
exports.lmUtils = lmUtils;
var rw_client_1 = require("./rw-client");
exports.RWLockClient = rw_client_1.RWLockClient;
exports.RWLockReadPrefClient = rw_client_1.RWLockReadPrefClient;
var rw_write_preferred_client_1 = require("./rw-write-preferred-client");
exports.RWLockWritePrefClient = rw_write_preferred_client_1.RWLockWritePrefClient;
var client_1 = require("./client");
exports.Client = client_1.Client;
exports.LMXClient = client_1.LMXClient;
exports.LvMtxClient = client_1.LvMtxClient;
var broker_1 = require("./broker");
exports.Broker = broker_1.Broker;
exports.LMXBroker = broker_1.LMXBroker;
exports.LvMtxBroker = broker_1.LvMtxBroker;
var shared_internal_1 = require("./shared-internal");
exports.LMXLockRequestError = shared_internal_1.LMXLockRequestError;
exports.LMXUnlockRequestError = shared_internal_1.LMXUnlockRequestError;
exports.r2gSmokeTest = function () {
    return true;
};
