// live-mutex/dist/shared-internal.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.forDebugging = {
    previousTime: Date.now()
};
var LMXLockRequestError;
(function (LMXLockRequestError) {
    LMXLockRequestError["GenericLockError"] = "generic_lock_error";
    LMXLockRequestError["MaxRetries"] = "max_retries";
    LMXLockRequestError["RequestTimeoutError"] = "request_timeout";
    LMXLockRequestError["BadArgumentsError"] = "bad_args";
    LMXLockRequestError["InternalError"] = "internal_error";
    LMXLockRequestError["UnknownException"] = "unknown_exception";
    LMXLockRequestError["WaitOptionSetToFalse"] = "wait_option_is_false";
})(LMXLockRequestError = exports.LMXLockRequestError || (exports.LMXLockRequestError = {}));
var LMXUnlockRequestError;
(function (LMXUnlockRequestError) {
    LMXUnlockRequestError["BadOrMismatchedId"] = "bad_or_mismatched_id";
    LMXUnlockRequestError["UnknownException"] = "unknown_exception";
    LMXUnlockRequestError["InternalError"] = "internal_error";
    LMXUnlockRequestError["GeneralUnlockError"] = "general_unlock_error";
})(LMXUnlockRequestError = exports.LMXUnlockRequestError || (exports.LMXUnlockRequestError = {}));
var RWStatus;
(function (RWStatus) {
    RWStatus["EndWrite"] = "end_write";
    RWStatus["BeginWrite"] = "begin_write";
    RWStatus["EndRead"] = "end_read";
    RWStatus["BeginRead"] = "begin_read";
    RWStatus["LockingWriteKey"] = "locking_write_key";
    RWStatus["UnlockingWriteKey"] = "unlocking_write_key";
})(RWStatus = exports.RWStatus || (exports.RWStatus = {}));
